### PR TITLE
[5/n][trainer] feat: flowgrpo trainer

### PR DESCRIPTION
**Pull request by zhtmike**
### What does this PR do?
Last piece of puzzle to make flowgrpo trainer runnable :) Following #5297, co-worked with @AndyZhou952 

Added:
- Adds FlowGRPO with a diffusion-oriented RL trainer and entrypoint.
- Groups diffusion algorithms and metric helpers under the diffusion trainer.
- Updates diffusion trainer, algorithm, and rollout configuration.
- Adds example FlowGRPO scripts and OCR-style image data preparation.
- Adds diffusion CPU tests, an E2E FlowGRPO diffusers run with dummy data.
- Applies small fixes / enhancements  across agent loops, workers, dataset loading, vLLM-omni rollout, etc.
- Adjusts the vLLM omni workflow, PR template, and sanity checks.

Documentations will be provided in next PR.
Script of full model RL / RL with non-collated reward model will be provided in the next PR.

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

Formal installation guide will be in next PR.
For now, you need to
1) install `vllm==0.18`
2) install `vllm-omni==0.18`
3) install this branch
4) prepare the dataset following `examples/flowgrpo_trainer/data_process/qwenimage_ocr.py`
5) run `examples/flowgrpo_trainer/run_qwen_image_ocr_lora.sh` to have a try.

This is wandb result running `examples/flowgrpo_trainer/run_qwen_image_ocr_lora.sh` with `trainer.val_before_train=True`

Val Score: 
螢幕截圖 2026-04-14 下午2 11 35

Critic:
螢幕截圖 2026-04-14 下午2 11 08

Actor:
螢幕截圖 2026-04-14 下午2 11 53

Performance
螢幕截圖 2026-04-14 下午2 12 09

Visualization
螢幕截圖 2026-04-14 下午2 12 33

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [x] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/trainer/diffusion/diffusion_algos.py [hunk 1]
```diff
@@ -13,15 +13,107 @@
...
+            batch_std = torch.std(scores)
+        else:
+            batch_std = None
+
+        bsz = scores.shape[0]
+        for i in range(bsz):
+            id2score[index[i]].append(scores[i])
+        for idx in id2score:
+            if len(id2score[idx]) == 1:
+                id2mean[idx] = torch.tensor(0.0)
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl/trainer/diffusion/diffusion_algos.py]: Fixed.

**Comment by zhtmike** [comment]
btw the current diffusion yaml is too clumsy and contains too much unnecessary llm configs, will refactor this after this one.

File: examples/flowgrpo_trainer/data_process/qwenimage_ocr.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    test_dataset = test_dataset.map(function=make_map_fn("test"), with_indices=True)
+
+    hdfs_dir = args.hdfs_dir
+    local_save_dir = args.local_dir
+    if local_save_dir is not None:
+        print("Warning: Argument 'local_dir' is deprecated. Please use 'local_save_dir' instead.")
+    else:
+        local_save_dir = args.local_save_dir
+
+    train_dataset.to_parquet(os.path.join(local_save_dir, "train.parquet"))
```
Comment (SamitHuang) [inline examples/flowgrpo_trainer/data_process/qwenimage_ocr.py:99]:
```suggestion
    local_save_dir = os.path.expanduser(local_save_dir)
    train_dataset.to_parquet(os.path.join(local_save_dir, "train.parquet"))
```

Comment (SamitHuang) [reply examples/flowgrpo_trainer/data_process/qwenimage_ocr.py:99]: The local_save_dir is not expanded using os.path.expanduser. Because the default value is "~/data/ocr", os.path.join(local_save_dir, "train.parquet") will attempt to write to a literal directory named ~ in the current working directory rather than the user's home folder

File: verl/trainer/config/rollout/diffusion_rollout.yaml [hunk 3]
```diff
@@ -23,6 +23,11 @@ width: 512
 # number of inference steps for diffusion model rollout
 num_inference_steps: 10
 
+# extra configs for algorithm specific features.
+# Model-specific diffusion sampling params (e.g. true_cfg_scale, guidance_scale,
+# max_sequence_length, noise_level)
+extra_configs: {}
```
Comment (SamitHuang) [inline verl/trainer/config/rollout/diffusion_rollout.yaml:29]: should we set `noise_level: 0.0` here to easier modification?

Comment (SamitHuang) [inline verl/trainer/config/rollout/diffusion_rollout.yaml:29]: btw, it's better to the **whole list** of arguments that are configured via `extra_configs`, either in comments and official docs

File: verl/trainer/config/rollout/diffusion_rollout.yaml [hunk 4]
```diff
@@ -38,14 +43,11 @@ val_kwargs:
...
   num_inference_steps: 40
 
-  # noise level for diffusion model rollout
-  noise_level: 0.0
-
   # random seed for validation
   seed: 42
+
+  # extra configs for algorithm specific features during validation.
+  extra_configs: {}
```
Comment (SamitHuang) [inline verl/trainer/config/rollout/diffusion_rollout.yaml:53]: same here. pls state what arguments can be set in `extra_config` of inference engine

Comment (zhtmike) [reply verl/trainer/config/rollout/diffusion_rollout.yaml:53]: Agree. extra_configs is a temporary design and will be dropped.

I prefer not to set the algorithm-specific configs (e.g., noise level, sde type, etc) or model-specific configs (maximum model length, true cfg scale, etc) into the general rollout config directly, which will make config bloated. So the extra_configs is a temp place to place these args here, which directly feeds these args into the rollout pipeline.

I am working on the config refactoring to make these things clearer. In the coming PR

File: verl/trainer/config/rollout/diffusion_rollout.yaml (hunk 3 above)
Comment (zhtmike) [reply verl/trainer/config/rollout/diffusion_rollout.yaml:29]: Forward-based RL algo like diffusionnft does not have this parameter. So it is not a universal parameter for rollout config

Comment (zhtmike) [reply verl/trainer/config/rollout/diffusion_rollout.yaml:29]: Agree. See below.

File: verl/workers/engine/fsdp/diffusers_impl.py [hunk 5]
```diff
@@ -779,9 +779,6 @@ def get_per_tensor_param(self, layered_summon=False, base_sync_done=False, **kwa
 
     @contextmanager
     def disable_adapter(self):
-        from diffusers.loaders import PeftAdapterMixin
-
-        assert isinstance(self.module, PeftAdapterMixin)
         try:
             self.module.disable_adapters()
             yield
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl/workers/engine/fsdp/diffusers_impl.py:784]: this is fine. FSDP automatically handle the wrapped attribute.

File: verl/trainer/diffusion/diffusion_algos.py [hunk 6]
```diff
@@ -13,15 +13,107 @@
...
+                    id2std[idx] = torch.std(scores_tensor)
+            else:
+                raise ValueError(f"no score in prompt index: {idx}")
+        for i in range(bsz):
+            if norm_adv_by_std_in_grpo:
+                scores[i] = (scores[i] - id2mean[index[i]]) / (id2std[index[i]] + epsilon)
+            else:
+                scores[i] = scores[i] - id2mean[index[i]]
+
+    return scores, scores
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl/trainer/diffusion/diffusion_algos.py:114]: we follow flowgrpo implementation

**Review by SamitHuang — APPROVED** [review]
LGTM. Pls consider improve the ease-of-use in the future PRs.
1) simplify the configuration and make them more structured, 
2) custom pipeline injection via python module paths may heavily leaks internal abstractions to the user. Maybe we can create an auto-resolver that can auto-gen the custome pipeline (QwenImagePipelineWithLogProb) based on the standard pipeline and target RL algo

**Comment by SamitHuang** [comment]
@wuxibin89 ready to merge, PTAL, thanks
